yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+   */
+  private val entityLock: AnyRef = new AnyRef
+
+  /**
+   * How long to wait for shutdown before giving up
+   */
+  private var maxTimeToWaitOnShutdown: Long = SHUTDOWN_WAIT_TIME
+
+  /** Domain ID for entities: may be null */
+  private var domainId: String = null
+
+  /** URI to timeline web application -valid after `serviceStart()` */
+  private var timelineWebappAddress: URI = _
+
+  /**
+   * Create a timeline client and start it. This does not update the
+   * `timelineClient` field, though it does verify that the field
+   * is unset.
+   *
+   * The method is private to the package so that tests can access it, which
+   * some of the mock tests do to override the timeline client creation.
+   * @return the timeline client
+   */
+  private [yarn] def createTimelineClient(): TimelineClient = {
+    require(timelineClient.isEmpty, "timeline client already set")
+    YarnTimelineUtils.createTimelineClient(sparkContext)
+  }
+
+  /**
+   * Get the timeline client.
+   * @return the client
+   * @throws Exception if the timeline client is not currently running
+   */
+  def getTimelineClient: TimelineClient = {
+    timelineClient.getOrElse(throw new Exception("Timeline client not running"))
+  }
+
+  /**
+   * Get the total number of processed events
+   * @return counter of events processed
+   */
+  def getEventsProcessed: Int = {
+    eventsProcessed.get
+  }
+
+  /**
+   * Get the total number of events queued
+   * @return the total event count
+   */
+  def getEventsQueued: Int = {
+    eventsQueued.get
+  }
+
+  /**
+   * Get the current size of the queue
+   * @return the current queue length
+   */
+  def getQueueSize: Int = {
+    actionQueue.size
+  }
+
+  /**
+   * Get the current batch size
+   * @return the batch size
+   */
+  def getBatchSize: Int = {
+    batchSize
+  }
+
+  /**
+   * Get the total number of failed posts events
+   * @return counter of timeline post operations which failed
+   */
+  def getEventPostFailures: Int = {
+    eventPostFailures.get
+  }
+
+  /**
+   * is the asynchronous posting thread active?
+   * @return true if the post thread has started; false if it has not yet/ever started, or
+   *         if it has finished.
+   */
+  def isPostThreadActive: Boolean = {
+    postThreadActive.get
+  }
+
+  /**
+   * The YARN application ID of this history service
+   * @return the application ID provided when the service started
+   */
+  def applicationId: ApplicationId = { _applicationId }
+
+  /**
+   * The YARN attempt ID of this history service
+   * @return the attempt ID provided when the service started
+   */
+  def attemptId: Option[ApplicationAttemptId] = { _attemptId }
+
+  /**
+   * Reset the timeline client
+   * <p>
+   * 1. Stop the timeline client service if running.
+   * 2. set the `timelineClient` field to `None`
+   */
+  def stopTimelineClient(): Unit = {
+    stopOptionalService(timelineClient)
+    timelineClient = None
+  }
+
+  /**
+   * Create the timeline domain
```

Comment:
It would be nice to have a better doc - like, explaining what the timeline domain is.